File: go/ie/se.go
package ie

import (
	"regexp"

	"github.com/matheuslm7/brazil-validator/go/internal/shared"
)

// Verified against the official SEFAZ-SE "Roteiro de Crítica da Inscrição
// Estadual" (sintegra.gov.br/Cad_Estados/cad_SE.html), including its worked
// example (27123456-3). Format: 8 digits + 1 check digit.
var (
	seAllowedChars = regexp.MustCompile(`^[\d\-\s]+$`)
	seWeights      = []int{9, 8, 7, 6, 5, 4, 3, 2}
)

func seNormalize(value string) string {
	return shared.RemoveNonDigits(value)
}

func seIsValid(value string) bool {
	if !seAllowedChars.MatchString(value) {
		return false
	}

	v := seNormalize(value)

	if len(v) != 9 {
		return false
	}

	return shared.Mod11CheckDigit(shared.WeightedSum(v, seWeights)) == int(v[8]-'0')
}

func seFormat(value string) string {
	v := seNormalize(value)

	if len(v) != 9 {
		return value
	}

	return v[:8] + "-" + v[8:9]
}

File: csharp/src/BrValidator/Ie/Se.cs
using System.Text.RegularExpressions;
using BrValidator.Internal;

namespace BrValidator.Ie
{
    /// <summary>
    /// SE: verified against the official SEFAZ-SE "Roteiro de Crítica da
    /// Inscrição Estadual" (sintegra.gov.br/Cad_Estados/cad_SE.html),
    /// including its worked example (27123456-3). Format: 8 digits + 1
    /// check digit.
    /// </summary>
    internal static class Se
    {
        private static readonly Regex AllowedChars = new Regex(@"^[\d\-\s]+$", RegexOptions.Compiled);
        private static readonly int[] Weights = { 9, 8, 7, 6, 5, 4, 3, 2 };

        public static string Normalize(string value) => Shared.RemoveNonDigits(value);

        public static bool IsValid(string value)
        {
            if (!AllowedChars.IsMatch(value))
            {
                return false;
            }

            string v = Normalize(value);

            if (v.Length != 9)
            {
                return false;
            }

            return Mod11.Mod11CheckDigit(Mod11.WeightedSum(v, Weights)) == (v[8] - '0');
        }

        public static string Format(string value)
        {
            string v = Normalize(value);

            if (v.Length != 9)
            {
                return value;
            }

            return $"{v.Substring(0, 8)}-{v.Substring(8, 1)}";
        }
    }
}
